### ISSUE TITLE
Nullable expression runs without a null check

### ISSUE THREAD

**Issue by tbm98** [flutter/flutter#88023]
doctor: 

<details>
<summary>Details</summary>

```
[✓] Flutter (Channel stable, 2.2.3, on macOS 11.4 20F71 darwin-x64, locale en)
    • Flutter version 2.2.3 at /Users/minhtran/fvm/versions/stable
    • Framework revision f4abaa0735 (6 weeks ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/minhtran/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at:
      /Library/Java/JavaVirtualMachines/jdk-11.0.12.jdk/Contents/Home/bin/java
    • Java version Java(TM) SE Runtime Environment 18.9 (build
      11.0.12+8-LTS-237)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5.1, Build version 12E507
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    ✗ Unable to find bundled Java version.
    • Try updating or re-installing Android Studio.

[✓] IntelliJ IDEA Community Edition (version 2021.2)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin version 59.0.4
    • Dart plugin version 212.4746.57

[✓] VS Code (version 1.59.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.25.0

[✓] Connected device (2 available)
    • sdk gphone x86 arm (mobile) • emulator-5554 • android-x86    • Android 11
      (API 30) (emulator)
    • Chrome (web)                • chrome        • web-javascript • Google
      Chrome 92.0.4515.131

! Doctor found issues in 1 category.
```

</details>

my code:
```dart
final LoyaltyProfile? result = _mapper?.from(response).copyWith(qrCode: request.customerQr);
    return result?.copyWith(
      hasPromotion: result.promotionsResources?.isNotEmpty == true,
      promotionsResources: result.promotionsResources
              ?.takeWhile(
                (value) => value.isApplicable,
              )
              .toList() ??
          [],
    );
```

linter show error but run fine without warning
<img width="905" alt="Screen Shot 2021-08-11 at 15 35 00" src="https://user-images.githubusercontent.com/52562340/128997264-15d98ea7-d01a-4121-9e6b-e545bffb0cb6.png">

when I modified it to `result!....` linter no longer show error but when I run, it shows a 
warning

<img width="842" alt="Screen Shot 2021-08-11 at 15 37 32" src="https://user-images.githubusercontent.com/52562340/128997690-a26e068a-bcf4-488a-bd2b-4c77dd4f00ec.png">
<img width="1316" alt="Screen Shot 2021-08-11 at 15 37 52" src="https://user-images.githubusercontent.com/52562340/128997721-c59d0758-c9a9-47ad-9917-23c19db99245.png">

**Comment by maheshmnj** [comment]
Hi @tbm98,
Thanks for filing the issue, please provide the output of `flutter doctor -v` along with [minimal reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example), and the complete error log (in the form of text), This helps to better investigate the issue.

Thank you.

**Comment by tbm98** [comment]
@maheshmnj [here you go](https://github.com/tbm98/flutter_app_bug_report/blob/main/lib/issue_pages/issue_88023.dart#L25)

but in the minimal project. after I fix the analysis, the console no longer show warning when I run

and `flutter doctor -v` already in the description

**Comment by maheshmnj** [comment]
@tbm98 Thanks for the details, I ran your sample code and I could reproduce the issue on the latest stable and the master channel. A nullable expression can be run without a null check.

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel stable, 2.2.3, on macOS 11.4 20F71 darwin-arm, locale en-GB)
    • Flutter version 2.2.3 at /Users/mahesh/Documents/flutter
    • Framework revision f4abaa0735 (6 weeks ago), 2021-07-01 12:46:11 -0700
    • Engine revision 241c87ad80
    • Dart version 2.13.4

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5.1, Build version 12E507
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] VS Code (version 1.58.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.25.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 11.4 20F71 darwin-arm
    • Chrome (web)    • chrome • web-javascript • Google Chrome 92.0.4515.131

• No issues found!
```

```
[✓] Flutter (Channel master, 2.5.0-6.0.pre.21, on macOS 11.4 20F71 darwin-arm, locale en-GB)
    • Flutter version 2.5.0-6.0.pre.21 at /Users/mahesh/Documents/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 40e8620a27 (4 days ago), 2021-08-06 09:14:50 +0430
    • Engine revision 431ac604da
    • Dart version 2.14.0 (build 2.14.0-383.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/mahesh/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • ANDROID_HOME = /Users/mahesh/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.5.1, Build version 12E507
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] VS Code (version 1.58.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.25.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 11.4 20F71 darwin-arm
    • Chrome (web)    • chrome • web-javascript • Google Chrome 92.0.4515.131

• No issues found!
```
</details>

**Comment by pq** [comment]
Hrmmm...

Maybe @stereotype441 has insight?

FWIW: as a general rule I think we tend to prefer early returns in cases like this:

```dart
final LoyaltyProfile? result = _mapper?.from(response).copyWith(qrCode: request.customerQr);
if (result == null) return [];

return result.copyWith(
      hasPromotion: result.promotionsResources?.isNotEmpty == true,
      promotionsResources: result.promotionsResources
              ?.takeWhile(
                (value) => value.isApplicable,
              )
              .toList() ??
          [],
    );
```

(Or maybe even earlier if `_mapper` is null.)

Or:

```dart
final LoyaltyProfile? result = _mapper?.from(response).copyWith(qrCode: request.customerQr);
if (result == null) return [];

final resources = result.promotionsResources;
if (resources == null) return [];

return result.copyWith(
      hasPromotion: resources.isNotEmpty,
      promotionsResources: resources.takeWhile(
                (value) => value.isApplicable,
              )
              .toList(),
    );

```

**Comment by stereotype441** [comment]
Thanks @pq for the ping, and thanks @tbm98 for the detailed repro.  This is happening due to a bug in the analyzer's implementation of null safety.  I've filed https://github.com/dart-lang/sdk/issues/46937 to track it, and should be able to get to fixing it next week.

In the meantime, @pq's suggested workarounds should all work.  Another possible workaround would be to change `result?.copyWith` to `result?.copyWith.call`, i.e.:

```dart
final LoyaltyProfile? result = _mapper?.from(response).copyWith(qrCode: request.customerQr);
    return result?.copyWith.call(
      hasPromotion: result.promotionsResources?.isNotEmpty == true,
      promotionsResources: result.promotionsResources
              ?.takeWhile(
                (value) => value.isApplicable,
              )
              .toList() ??
          [],
    );
```
I hope that helps!

**Comment by danagbemava-nc** [comment]
It looks like this issue has since been fixed. 
I checked on `flutter 2.5.3` where the issue reproduced but it didn't reproduce on `flutter 2.8.1`.

@tbm98 can you confirm if this has been fixed for you as well? 

<details>
<summary>screenshots</summary>

| 2.5.3 | 2.8.1 |
| -- | -- |
| <img width="1136" alt="Screenshot 2022-01-20 at 11 04 06" src="https://user-images.githubusercontent.com/88313112/150327374-874569b4-e175-4243-86f4-efefe37229b8.png"> |<img width="1136" alt="Screenshot 2022-01-20 at 11 04 30" src="https://user-images.githubusercontent.com/88313112/150327380-64b20b5e-43c7-4104-9017-3dbb19f58099.png"> |

</details>

**Comment by tbm98** [comment]
I will try and tell you soon

**Comment by tbm98** [comment]
Yes. this issue has been fixed on 2.8.1. Thanks!
